compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java

```diff
@@ -846,9 +851,23 @@ private BType createRecordType(BLangExpression expr, BConstantSymbol constantSym
         }
 
         createTypeDefinition(recordType, pos, env);
-        BIntersectionType intersectionType = ImmutableTypeCloner.getImmutableIntersectionType(pos, types,
-                recordType, env, symTable, anonymousModelHelper, names, new HashSet<>());
-        return intersectionType;
+        updateRecordFields(recordType, pos, env);
+        recordType.tsymbol.flags |= Flags.READONLY;
+        recordType.flags |= Flags.READONLY;
+        return recordType;
+    }
+
+    private void updateRecordFields(BRecordType recordType, Location pos, SymbolEnv env) {
+        BTypeSymbol structureSymbol = recordType.tsymbol;
+        for (BField field : recordType.fields.values()) {
+            field.type = ImmutableTypeCloner.getImmutableType(pos, types, field.type, env,
```

Comment:
Within this we still create type defs, right? E.g., mapping within mapping. Shall we create an issue to avoid this also as a future improvement?